jason/jason-dummy/src/utils/closure.rs

```diff
@@ -0,0 +1,140 @@
+
+/// Dart function used to invoke other Dart closures without arguments.
+static mut NO_ARGS_FN_CALLER: Option<UnitArgFnCaller> = None;
+
+/// Dart function used to invoke other Dart closures that accept an `i64`
+/// argument.
+static mut INT_ARG_FN_CALLER: Option<IntArgFnCaller> = None;
+
+/// Registers the provided [`PointerArgFnCaller`] as [`PTR_ARG_FN_CALLER`]. Must
+/// be called by dart during FFI initialization.
+#[no_mangle]
+pub unsafe extern "C" fn register_ptr_arg_fn_caller(
+    caller: PointerArgFnCaller,
+) {
+    PTR_ARG_FN_CALLER = Some(caller);
+}
+
+/// Registers the provided [`UnitArgFnCaller`] as [`NO_ARGS_FN_CALLER`]. Must be
+/// called by dart during FFI initialization.
+#[no_mangle]
+pub unsafe extern "C" fn register_no_args_fn_caller(f: UnitArgFnCaller) {
+    NO_ARGS_FN_CALLER = Some(f)
+}
+
+/// Registers the provided [`IntArgFnCaller`] as [`INT_ARG_FN_CALLER`]. Must be
+/// called by dart during FFI initialization.
+#[no_mangle]
+pub unsafe extern "C" fn register_int_arg_fn_caller(f: IntArgFnCaller) {
+    INT_ARG_FN_CALLER = Some(f);
+}
+
+// TODO: Print exception if Dart closure throws.
+/// Dart closure that can be called from Rust.
+pub struct DartClosure<T> {
+    /// [`Dart_PersistentHandle`] to the Dart closure that should be called.
+    cb: Dart_PersistentHandle,
+    _arg_kind: PhantomData<T>,
+}
+
+impl<T> DartClosure<T> {
+    /// Creates new [`DartClosure`] from the provided [`Dart_Handle`] to the
+    /// Dart closure. Persists the provided [`Dart_Handle`] so it wont be moved
+    /// by the Dart VM GC.
+    pub fn new(cb: Dart_Handle) -> Self {
+        Self {
+            cb: unsafe { Dart_NewPersistentHandle_DL_Trampolined(cb) },
+            _arg_kind: PhantomData::default(),
+        }
+    }
+}
+
+impl DartClosure<()> {
+    /// Calls underlying Dart closure.
+    pub fn call0(&self) {
+        unsafe {
+            let fn_handle = Dart_HandleFromPersistent_DL_Trampolined(self.cb);
+            NO_ARGS_FN_CALLER.unwrap()(fn_handle);
+        }
+    }
+}
+
+impl<T: ForeignClass> DartClosure<T> {
+    /// Calls underlying Dart closure with provided [`ForeignClass`] argument.
+    pub fn call1(&self, arg: T) {
+        unsafe {
+            let fn_handle = Dart_HandleFromPersistent_DL_Trampolined(self.cb);
+            PTR_ARG_FN_CALLER.unwrap()(fn_handle, arg.into_ptr().cast::<()>());
+        }
+    }
+}
+
+/// Implements [`DartClosure::call`] casting argument to `i64`. Should be
+/// called for all integer types that fit into 2^63.
+macro_rules! impl_dart_closure_for_int {
+    ($arg:ty) => {
+        impl DartClosure<$arg> {
+            /// Calls underlying Dart closure with provided argument.
+            pub fn call1(&self, arg: $arg) {
+                unsafe {
+                    let fn_handle =
+                        Dart_HandleFromPersistent_DL_Trampolined(self.cb);
+                    INT_ARG_FN_CALLER.unwrap()(fn_handle, arg as i64);
```

Comment:
It seems that a regular generic implementation goes better here:

```rust
impl<T> DartClosure<T> where i64: From<T> {
    /// Calls underlying Dart closure with provided [`ForeignClass`] argument.
    pub fn call1(&self, arg: T) {
        unsafe {
            let fn_handle = Dart_HandleFromPersistent_DL_Trampolined(self.cb);
            INT_ARG_FN_CALLER.unwrap()(fn_handle, arg.into());
        }
    }
}
```

If required, or orphan rules are pain-in-ass, the types set can be restricted further with an add-hoc marker trait.

Just no need to use macros where a regular monomorphization plays well.